data-science-onramp/data-ingestion/ingestion_test.py

```diff
@@ -73,25 +74,38 @@
 
 @pytest.fixture(autouse=True)
 def setup_and_teardown_cluster():
-    # Create cluster using cluster client
-    cluster_client = dataproc.ClusterControllerClient(
-        client_options={"api_endpoint": f"{CLUSTER_REGION}-dataproc.googleapis.com:443"}
-    )
+    try:
+        # Create cluster using cluster client
+        cluster_client = dataproc.ClusterControllerClient(
+            client_options={
+                "api_endpoint": f"{CLUSTER_REGION}-dataproc.googleapis.com:443"
+            }
+        )
 
-    operation = cluster_client.create_cluster(
-        project_id=PROJECT_ID, region=CLUSTER_REGION, cluster=CLUSTER_CONFIG
-    )
+        operation = cluster_client.create_cluster(
+            project_id=PROJECT_ID, region=CLUSTER_REGION, cluster=CLUSTER_CONFIG
+        )
 
-    # Wait for cluster to provision
-    operation.result()
+        # Wait for cluster to provision
+        operation.result()
+    except InternalServerError as e:
+        # Clean up leftovers if there is an error creating the fixture
+        operation = cluster_client.delete_cluster(
+            project_id=PROJECT_ID, region=CLUSTER_REGION, cluster_name=DATAPROC_CLUSTER
+        )
+        operation.result()
+        raise e
 
     yield
 
-    # Delete cluster
-    operation = cluster_client.delete_cluster(
-        project_id=PROJECT_ID, region=CLUSTER_REGION, cluster_name=DATAPROC_CLUSTER
-    )
-    operation.result()
+    try:
+        # Delete cluster
+        operation = cluster_client.delete_cluster(
+            project_id=PROJECT_ID, region=CLUSTER_REGION, cluster_name=DATAPROC_CLUSTER
+        )
+        operation.result()
+    except NotFound:
+        print("Cluster already deleted")
```

Comment:
Thoughts on moving `yield` to inside the `try:` block and using `finally` instead?

```python
        # Wait for cluster to provision
        operation.result()
        yield
    finally:
        # Clean up leftovers if there is an error creating the fixture
        try:
            operation = cluster_client.delete_cluster(
                project_id=PROJECT_ID, region=CLUSTER_REGION, cluster_name=DATAPROC_CLUSTER
            )
            operation.result()
        except NotFound:
            print("Cluster already deleted")
```

Reply:
great idea! sometimes I forget about finally. Thanks for the reminder - I'll make that change now!